sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -63,55 +64,26 @@ private[sql] object OrcFilters extends OrcFiltersBase {
    */
   def createFilter(schema: StructType, filters: Seq[Filter]): Option[SearchArgument] = {
     val dataTypeMap = schema.map(f => f.name -> f.dataType).toMap
+    val orcFilterConverter = new OrcFilterConverter(dataTypeMap)
     for {
       // Combines all convertible filters using `And` to produce a single conjunction
-      conjunction <- buildTree(convertibleFilters(schema, dataTypeMap, filters))
+      conjunction <- buildTree(filters.flatMap(orcFilterConverter.trimUnconvertibleFilters))
       // Then tries to build a single ORC `SearchArgument` for the conjunction predicate
-      builder <- buildSearchArgument(dataTypeMap, conjunction, newBuilder)
+      builder <- orcFilterConverter.buildSearchArgument(conjunction, newBuilder)
```

Comment:
`buildSearchArgument` already trimmed filters, so we don't need the for loop here?

Reply:
I'm not sure which loop you're referring to here. The `for-yield` is used for dealing with the `Option`s that both of these methods can return, and they can still return Options so we still need the for loop.

We can create a separate function which takes in a trimmed filter and just handles the building, but I'm not sure how useful that would be. I kind of like the isolation and slight guarantees that `build` will only be called on a trimmed filter that having one common `buildSearchArgument` method gives us.

Let me know if you had something else in mind.

Reply:
The code here:
```
for {
  conjunction <- buildTree(filters.flatMap(orcFilterConverter.trimUnconvertibleFilters))
  builder <- orcFilterConverter.buildSearchArgument(conjunction, newBuilder)
}
```
However, `buildSearchArgument` calls `performFilter` which is same as calling `trimUnconvertibleFilters`. That said, we call `trimUnconvertibleFilters` twice.

Reply:
Got rid of the explicit `trimUnconvertibleFilters` call in this method.

We still need the for-yield because both `buildTree` and `buildSearchArgument` return Options.